manta-accounting/src/transfer/mod.rs

```diff
@@ -0,0 +1,2371 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Transfer Protocol
+
+use crate::asset::{Asset, AssetId, AssetValue};
+use alloc::vec::Vec;
+use core::{fmt::Debug, hash::Hash, iter, marker::PhantomData};
+use manta_crypto::{
+    accumulator::{Accumulator, MembershipProof, Model},
+    commitment::CommitmentScheme,
+    constraint::{
+        Add, Allocator, Constant, ConstraintSystem, Derived, Equal, ProofSystem, ProofSystemInput,
+        Public, Secret, ValueSource, Variable,
+    },
+    encryption::hybrid::{DecryptedMessage, EncryptedMessage, HybridPublicKeyEncryptionScheme},
+    hash::BinaryHashFunction,
+    key::KeyAgreementScheme,
+    rand::{CryptoRng, RngCore, Sample},
+};
+
+#[cfg(feature = "serde")]
+use manta_util::serde::{Deserialize, Serialize};
+
+pub mod batch;
+pub mod canonical;
+
+#[cfg(feature = "test")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "test")))]
+pub mod test;
+
+#[doc(inline)]
+pub use canonical::Shape;
+
+/// Returns `true` if the [`Transfer`] with this shape would have public participants.
+#[inline]
+pub const fn has_public_participants(sources: usize, sinks: usize) -> bool {
+    (sources + sinks) > 0
+}
+
+/// Transfer Configuration
+pub trait Configuration {
+    /// Secret Key Type
+    type SecretKey: Clone + Sample;
+
+    /// Public Key Type
+    type PublicKey: Clone;
+
+    /// Key Agreement Scheme Type
```

Comment:
Most of the doc strings are a repeat of the variable names. It might be helpful to give some short descriptions here for at least the non-obvious ones.

Reply:
Not sure which ones are non-obvious here.